File: PhaseJump/Xcode/PhaseJump/PhaseJumpTests/G Tests/Core/ValveTests.cpp
#include "gtest/gtest.h"
#include "Valve.h"

using namespace PJ;
using namespace std;

namespace ValveTests {
    class TestValve : public Valve
    {
    public:
        TestValve(bool isOn, float& progress) : Valve(isOn) {
            SetOnValveUpdateFunc([&](auto& valve) {
                progress = valve.Value();
            });
        }
    };
}

using namespace ValveTests;

TEST(Valve, Init)
{
    {
        float progress{};
        TestValve sut(true, progress);
        EXPECT_TRUE(sut.IsOn());
        EXPECT_EQ(1, progress);
    }
    {
        float progress{};
        TestValve sut(false, progress);
        EXPECT_FALSE(sut.IsOn());
        EXPECT_EQ(0, progress);
    }
}

TEST(Valve, TurnOnWhenOff)
{
    float progress{};
    TestValve sut(false, progress);
    EXPECT_FALSE(sut.IsOn());

    sut.TurnOn(2);
    EXPECT_EQ(Valve::StateType::TurnOn, sut.State());
    EXPECT_EQ(0, sut.Value());

    sut.OnUpdate({1});
    EXPECT_EQ(0.5f, sut.Value());
    EXPECT_EQ(0.5f, progress);

    sut.OnUpdate({1});
    EXPECT_EQ(1.0f, sut.Value());
    EXPECT_EQ(1.0f, progress);

    EXPECT_TRUE(sut.IsOn());
}

TEST(Valve, TurnOnInterruptTurnOff)
{
    Valve sut(false);
    EXPECT_FALSE(sut.IsOn());

    sut.TurnOn(2);
    EXPECT_EQ(Valve::StateType::TurnOn, sut.State());
    EXPECT_EQ(0, sut.Value());

    sut.OnUpdate({.5f});
    EXPECT_EQ(0.25f, sut.Value());

    sut.TurnOff(2);
    EXPECT_EQ(Valve::StateType::TurnOff, sut.State());
    EXPECT_EQ(0.25f, sut.Value());

    sut.OnUpdate({.5f});
    EXPECT_EQ(0, sut.Value());
    EXPECT_TRUE(sut.IsOff());
}

TEST(Valve, TurnOnWhenOn)
{
    Valve sut(true);
    EXPECT_TRUE(sut.IsOn());

    sut.TurnOn(2);
    EXPECT_EQ(1.0f, sut.Value());
    EXPECT_TRUE(sut.IsOn());
}

TEST(Valve, TurnOffWhenOn)
{
    Valve sut(true);
    EXPECT_FALSE(sut.IsOff());

    sut.TurnOff(2);
    EXPECT_EQ(Valve::StateType::TurnOff, sut.State());
    EXPECT_EQ(1, sut.Value());

    sut.OnUpdate({0.5f});
    EXPECT_EQ(0.75f, sut.Value());

    sut.OnUpdate({1.5f});
    EXPECT_EQ(0, sut.Value());

    EXPECT_TRUE(sut.IsOff());
}

TEST(Valve, TurnOffWhenOff)
{
    Valve sut(false);
    EXPECT_TRUE(sut.IsOff());

    sut.TurnOff(2);
    EXPECT_EQ(0, sut.Value());
    EXPECT_TRUE(sut.IsOff());
}

TEST(Valve, TurnOffInterruptTurnOn)
{
    Valve sut(true);
    EXPECT_TRUE(sut.IsOn());

    sut.TurnOff(2);
    EXPECT_EQ(Valve::StateType::TurnOff, sut.State());
    EXPECT_EQ(1, sut.Value());

    sut.OnUpdate({.5f});
    EXPECT_EQ(0.75f, sut.Value());

    sut.TurnOn(2);
    EXPECT_EQ(Valve::StateType::TurnOn, sut.State());
    EXPECT_EQ(0.75f, sut.Value());

    sut.OnUpdate({.5f});
    EXPECT_EQ(1, sut.Value());
    EXPECT_TRUE(sut.IsOn());
}

TEST(Valve, TurnOnImmediate)
{
    float progress{};
    TestValve sut(false, progress);
    EXPECT_FALSE(sut.IsOn());

    sut.TurnOn(0);
    EXPECT_EQ(Valve::StateType::On, sut.State());
    EXPECT_EQ(1, sut.Value());
    EXPECT_EQ(1, progress);
}

TEST(Valve, TurnOffImmediate)
{
    float progress{};
    TestValve sut(true, progress);
    EXPECT_TRUE(sut.IsOn());

    sut.TurnOff(0);
    EXPECT_EQ(Valve::StateType::Off, sut.State());
    EXPECT_EQ(0, sut.Value());
    EXPECT_EQ(0, progress);
}

File: Experimental/Swift/PhaseJumpPro/Tests/PhaseJumpProTests/ValveTests.swift
import Foundation
import Testing
@testable import PhaseJumpPro

struct ValveTests {
    class Counter {
        var progress = 0.0
    }
    
    func makeTestValve(counter: Counter, isOn: Bool) -> Valve {
        var result = Valve(isOn: isOn)
        result.onValveUpdateFunc = {
            counter.progress = $0.value
        }
        
        return result
    }

    @Test
    func initializeIsOn() {
        let counter = Counter()
        let sut = makeTestValve(counter: counter, isOn: true)
        #expect(sut.isOn)
        #expect(1 == counter.progress)
    }

    @Test
    func initializeIsOff() {
        let counter = Counter()
        let sut = makeTestValve(counter: counter, isOn: false)
        #expect(false == sut.isOn)
        #expect(0 == counter.progress)
    }
    
    @Test
    func turnOnWhenOff() {
        let counter = Counter()
        let sut = makeTestValve(counter: counter, isOn: false)
        #expect(false == sut.isOn)

        sut.turnOn(duration: 2)
        #expect(.turnOn == sut.state)
        #expect(0 == sut.value)

        sut.onUpdate(time: .init(delta: 1))
        #expect(0.5 == sut.value)
        #expect(0.5 == counter.progress)

        sut.onUpdate(time: .init(delta: 1))
        #expect(1.0 == sut.value)
        #expect(1.0 == counter.progress)

        #expect(sut.isOn)
    }

    @Test
    func turnOnInterruptTurnOff() {
        let sut = Valve(isOn: false)
        #expect(false == sut.isOn)

        sut.turnOn(duration: 2)
        #expect(.turnOn == sut.state)
        #expect(0 == sut.value)

        sut.onUpdate(time: .init(delta: 0.5))
        #expect(0.25 == sut.value)

        sut.turnOff(duration: 2)
        #expect(.turnOff == sut.state)
        #expect(0.25 == sut.value)

        sut.onUpdate(time: .init(delta: 0.5))
        #expect(0 == sut.value)
        #expect(sut.isOff)
    }

    @Test
    func turnOnWhenOn() {
        let sut = Valve(isOn: true)
        #expect(sut.isOn)

        sut.turnOn(duration: 2)
        #expect(1.0 == sut.value)
        #expect(sut.isOn)
    }

    @Test
    func turnOffWhenOn() {
        let sut = Valve(isOn: true)
        #expect(false == sut.isOff)

        sut.turnOff(duration: 2)
        #expect(.turnOff == sut.state)
        #expect(1 == sut.value)

        sut.onUpdate(time: .init(delta: 0.5))
        #expect(0.75 == sut.value)

        sut.onUpdate(time: .init(delta: 1.5))
        #expect(0 == sut.value)

        #expect(sut.isOff)
    }

    @Test
    func turnOffWhenOff() {
        let sut = Valve(isOn: false)
        #expect(sut.isOff)

        sut.turnOff(duration: 2)
        #expect(0 == sut.value)
        #expect(sut.isOff)
    }
    
    @Test
    func turnOffInterruptTurnOn() {
        let sut = Valve(isOn: true)
        #expect(sut.isOn)

        sut.turnOff(duration: 2)
        #expect(.turnOff == sut.state)
        #expect(1 == sut.value)

        sut.onUpdate(time: .init(delta: 0.5))
        #expect(0.75 == sut.value)

        sut.turnOn(duration: 2)
        #expect(.turnOn == sut.state)
        #expect(0.75 == sut.value)

        sut.onUpdate(time: .init(delta: 0.5))
        #expect(1 == sut.value)
        #expect(sut.isOn)
    }

    @Test
    func turnOnImmediate() {
        let counter = Counter()
        let sut = makeTestValve(counter: counter, isOn: false)
        #expect(false == sut.isOn)

        sut.turnOn(duration: 0)
        #expect(.on == sut.state)
        #expect(1 == sut.value)
        #expect(1 == counter.progress)
    }
    
    @Test
    func turnOffImmediate() {
        let counter = Counter()
        let sut = makeTestValve(counter: counter, isOn: true)
        #expect(sut.isOn)

        sut.turnOff(duration: 0)
        #expect(.off == sut.state)
        #expect(0 == sut.value)
        #expect(0 == counter.progress)
    }
}
